계산기 [STEP 1] Kobe

devKobe24 opened the pull request:
안녕하세요 원비(@wonbi92) 🙌
처음으로 원비께 리뷰를 받게된 코비🥴 입니다.
잘 부탁드립니다.

---

## 🏗️ UML
<img src = "https://github.com/devKobe24/images/blob/main/QueueUML_2.drawio.png?raw=true" width = 550>

---
## 🤔 고민한 점
1. Queue의 원리에 대해서 고민했습니다.
- Queue라는 자료구조가 어떤 원리와 어떤 구조로 생겼는지 찾아보았으며 많은 변형이 있다는 것을 알게 되었습니다.

2. 배열 vs 단방향 리스트 
- Queue 구조를 구현시 배열과 단방향 리스트로 구현할 수 있다는 것을 알게되었습니다. 아직 단방향 리스트에 대한 정확한 개념과 단방향 리스트로 구현할 수 있는 충분한 학습이 되어있지 않아 배열로 구현하였습니다.
- 배열과 단방향 리스트로 큐를 구현했을 때의 차이점은 시간 복잡도에 있다는 것을 알았습니다. 예를 들어 배열의 첫 번째 요소가 빠지면 뒤에 있는 요소들이 앞으로 한 칸씩 자리를 다시 정렬해야해서 O(n)이 걸리지만 단방향 리스트의 경우 첫 번째 요소가 빠진다고 해도 노드를 가리키고 있는 참조만 끊어주면 되기 때문에 O(1)이 걸린다고 학습했습니다.

> 이와 같은 이유로 배열로 큐를 구현하게 되었습니다.

---

## 🙋‍♂️ 조언을 얻고 싶은 부분
1. TDD.
- 함수가 구현되어 있지 않은 상태에서 어떻게 테스트를 할지 계속해서 고민이 됩니다 그래서 제가 선택한 방법은 일단 빈 함수를 만들고 [RED] -> 내부에 함수 로직 구성 [GREEN] -> 리팩토링 [BLUE] 이렇게 하고 있는데 이 방식이 맞는지 궁굼합니다.

2. Queue의 단방향 리스트 구현.
- Queue를 단방향 리스트 구조를 이용해서 구현해야 하는 경우 어떤 점을 중점적으로 고민하고 구현해야할까요?

Commit:
refactor: MVC 형태로 폴더 구성 정리 및 파일 이동.

Commit:
feat: CalculatorItemQueue 파일 생성, Node 파일 생성 및 클래스 생성, CalculatorTestsPlan 생성, CalculatorItemQueueTests 파일 생성, CalculatorNodeTests 파일 생성.

Commit:
feat: Node 클래스 생성 테스트 구현.

Commit:
test: Node 클래스 value 프로퍼티에 제네릭 타입이 들어가는 테스트 추가.

Commit:
refactor: Node 파일 삭제, CalculatorNodeTests 파일 삭제, CalculatorTestsPlan 수정.

Commit:
feat: 배열에 숫자를 담는 함수 생성.

Commit:
test: 배열에 값이 들어가는것을 확인하기 위한 테스트 생성 및 확인.

Commit:
chore: enqueue로 함수명 수정.

Commit:
test: 배열에서 값이 빠지는 기능 테스트 [RED]

Commit:
refactor: enqueue에 반환값 생성

Commit:
test: 배열에서 값이 빠지는 기능 테스트 [GREEN]

Commit:
test: 배열에 요소가 없을때 peek 함수를 호출하면 nil을 반환한다 테스트[RED]

Commit:
test: 배열에 요소가 없을 때 peek 함수를 호출하면 nil을 반환한다 [GREEN]

Commit:
test: clear 함수를 호출시 배열의 요소가 모두 비워졌을때 true를 반환한다 테스트[RED]

Commit:
test: clear 함수를 호출시 배열의 요소가 모두 비워졌을때 true를 반환한다[GREEN]

Commit:
refactor: clear 함수 수정 [BLUE]

wonbi92 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,36 @@
+
+protocol CalculateItem {
+	
+}
```
빈 프로토콜을 선언할 때는 대괄호 사이를 한칸 띄어쓰는게 일반적인 방법입니다. 정답은 아니니 참고하세용~
```suggestion
protocol CalculateItem { }
```

devKobe24 replied:
네!! 꼭 참고하겠습니다 🙋‍♂️

wonbi92 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,36 @@
+	
+}
+
+struct CalculatorItemQueue: CalculateItem {
```
`CalculatorItemQueue`가 `CalculateItem`을 채택하고 있군요..! 요구사항을 보면 
> `CalculatorItemQueue`에서 **다루는 요소**는 `CalculateItem` 프로토콜을 준수합니다.

라고 하고 있어요! 지금 코드대로 라면 "다루는 요소"가 `CalculateItem` 프로토콜을 준수한다고 할 수 없을거 같은데.. 🤔

devKobe24 replied:
요구사항을 조금 더 자세히 보고 이해한 후 수정해보도록 하겠습니다. 🙌

wonbi92 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,36 @@
+struct CalculatorItemQueue: CalculateItem {
+	private var elements: [Int] = []
+	
+	mutating func enqueue(_ value: Int) -> [Int] {
+		elements.append(value)
+		return elements
+	}
```
`enqueue`하는 메서드가 배열을 반환하는 이유를 알 수 있을까요? 👀

devKobe24 replied:
TDD시 테스트를 위해 반환값을 주었습니다. 반환값이 굳이 필요 없는 함수에 TDD를 위해 반환값을 준 것을 수정해보도록 하겠습니다 🙌

wonbi92 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,36 @@
+		return elements.removeFirst()
+	}
+	
+	func peek() -> Int? {
+		return elements.first
+	}
```
요 메서드는 어디서 사용하는 메서드일까요? 👀

devKobe24 replied:
처음 값이 0이나 기호가 안들어가게 만들기 위해 큐의 첫 번째 요소를 검사하려고 만든 메서드입니다.🙋‍♂️

wonbi92 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,36 @@
+		return elements.first
+	}
+	
+	mutating func clear() -> Bool {
+		elements.removeAll()
+		if elements == [] {
+			return true
+		}
+		return false
+	}
```
`removeAll()`을 호출한 후에 배열이 비어있는지 확인하는 로직이군요..! 다만, `removeAll()`메서드가 실패할 수 있는 메서드일까요? 🤔 
제 생각에는 이 로직에서는 무조건 if문을 타고 true를 반환할거 같아요! 배열이 비어있는지 확인하는 로직을 추가하신 이유를 알 수 있을까요? 😅

devKobe24 replied:
지금 보니 if 문이 없으면 더 깔끔한 코드가 될 것 같습니다. 더 깊이 생각해보고 수정해보겠습니다 🙋‍♂️

wonbi92 commented on Calculator/CalculatorItemQueueTests/CalculatorItemQueueTests.swift:
```diff
@@ -0,0 +1,91 @@
+	
+	func tests_peek함수를_호출시_배열의_요소가_여러개여도_첫번째_인덱스를_반환한다() {
+		// given
+		var input = sut.enqueue(1)
```
`input`변수에 대한 컴파일 경고가 뜨고 있네요! 이 컴파일에러는 무엇이고, 어떻게 해결할 수 있을까요?

devKobe24 replied:
변수가 쓰여졌지만 사용은 안되서 나는 에러입니다. 와일드 카드 패턴을 사용해서 해결할 수 있을 것 같습니다 🙋‍♂️

wonbi92 commented on Calculator/CalculatorItemQueueTests/CalculatorItemQueueTests.swift:
```diff
@@ -0,0 +1,91 @@
+		sut = nil
+    }
+	
+	func tests_배열에_값이_들어간다() {
```
테스트 케이스를 작성하는 방법은 사바사, 회바회기 때문에 정답은 없습니다만, 각 테스트 케이스별 표현 방식은 통일 시켜주시는게 좋겠죠? 또한, 케이스를 작성하실 때는 어떤 케이스를 테스트하는지 명확하게 전달해줄 필요가 있습니다. 

현재 작성하신 케이스는 변인통제를 제대로 명시하지 않아 테스트 케이스를 해석하기 모호해지는 것 같습니다.
배열에 어떤 값이 들어가는지, 얼마나 들어가는지 모르고, 배열에 값이 제대로 들어가는지 확인하는 테스트인지, 배열에 값이 들어가는 로직 자체를 확인하는 테스트인지.. 등등 모호한 부분이 너무 많습니다. 좀 더 정확한 표현으로 이 테스트가 어떤 테스트인지 명시해주세요!

devKobe24 replied:
테스트 케이스 변인통제를 제대로 명시하여 테스트 케이스를 해석하기에 모호해지는 일이 없도록 수정하고 이후에는 테스트 케이스 변인통제를 제대로 명확하게 명시하도록 하겠습니다 🙋‍♂️

wonbi92 reviewed (changes requested):
@devKobe24 
반갑습니다 코비! 3주동안 계산기 리뷰를 맡게된 Wonbi입니다! 앞으로 잘 부탁드려요~ 😁

step1진행하시느라 고생하셨습니다!

몇가지 수정사항과 참고하실만한 부분들을 코맨트 드렸으니 쭉 읽어보시고 궁금하신 점 있으시면 말씀해주세요~ 💪

### 아래는 질문주신 부분에 대한 제 생각입니다! 😁
---
> 함수가 구현되어 있지 않은 상태에서 어떻게 테스트를 할지 계속해서 고민이 됩니다 그래서 제가 선택한 방법은 일단 빈 함수를 만들고 [RED] -> 내부에 함수 로직 구성 [GREEN] -> 리팩토링 [BLUE] 이렇게 하고 있는데 이 방식이 맞는지 궁굼합니다.
- TDD는 방법론이기 때문에 개발 방식 외에 모든 결정은 개발자 스스로 결정해야 합니다. 코비가 고민하시고 결정하신 방법이 맞다 틀리다 라고 할 수는 없을거 같아요!
- 말씀하시는 빈 함수가 어떤 함수인지 파악하기가 조금 힘드네요! 테스트 하고자 하시는 객체의 메서드인지 테스트 케이스인지.. 
- 일반적으로 TDD는 테스트를 작성한 후 그 테스트를 성공시키기 위해 리팩토링을 진행하죠! 그런데 테스트를 작성하려면 테스트 하려는 로직 자체는 필요하잖아요? 그렇기 때문에, 테스트 하려는 객체에 비어있는 함수를 구현하고, 이 함수를 통해 원하는 결과가 나오는지 테스트하는 테스트 케이스를 작성하신 후, 빈 함수를 리팩토링하는 방식으로 진행될거 같아요!

Commit:
:adhesive_bandage: Fix: CalculatorItemQueue 코드 수정, CalculatorItemQueueTests 테스트 삭제.

CalculatorItemQueue 내부에 있는 코드들을 요구 사항에 맞게 수정해봤습니다. 또한 새롭게 테스트를 하기 위해 CalculatorItemQueueTests 테스트를 모두 삭제했습니다.

Commit:
:sparkles: Feat: elements 요소를 가져오기 위한 함수 생성.

테스트를 위해 private var elements를 접근하고 가져오기 위해 함수를 작성하였습니다.

Commit:
:white_check_mark: Test: enqueue 함수 호출시 배열에 요소가 담긴다 [GREEN]

enqueue 함수 호출시 배열에 요소가 담기는지 확인하기 위한 테스트를 작성하고 테스트 한 결과 성공하였습니다.

Commit:
:white_check_mark: Test: enqueue 여러번 호출시 테스트 성공[GREEN]

enqueue 함수를 여러번 호출 했을 경우 elements 배열에 요소들이 쌓이는지 확인하기 위한 테스트를 시행하였고 성공하였습니다.

Commit:
:test_tube: Test: 빈 배열에 dequeue시 nil 반환 테스트 [RED]

배열이 비어있는 상태에서 dequeue를 호출했을 경우 nil을 반환하는지 테스트 했지만 실패했습니다.

Commit:
:adhesive_bandage: Fix: where 구문 추가, Equatable 채택, dequeue 로직 변경.

테스트 중 비교 연산을 하려 시도했습니다 그러려면 Equatable이 필요해 where 구문으로 Equatable을 추가로 채택해 주었습니다. 또한 dequeue 내부에서 removeFirst()를 실행시 오류가 나는 부분이있어 내부 로직을 수정했습니다.

Commit:
:white_check_mark: Test: dequeue 호출시 배열에서 요소가 하나씩 빠짐[GREEN]

배열에 요소가 3개 들어있게 만든 다음 첫 번째 요소가 빠지게 만드는 테스트를 한 결과 성공적으로 빠지는 것을 확인했습니다.

Commit:
:white_check_mark: Test: peek 함수 호출시 nil 반환.

빈 배열에 peek 함수를 호출할 경우 nil을 반환하는지 테스트를 해본 결과 nil을 반환하는 것을 확인했습니다.

Commit:
:white_check_mark: Test: peek 함수 호출시 첫 번째 요소를 가져온다.

배열에 요소 [1,2,3]이 있을 경우 첫 번째 요소인 1을 peek을 호출해서 가져올 수 있는지 테스트 해봤습니다 그 결과 성공했습니다.

Commit:
:white_check_mark: Test: clear 함수 호출 테스트 [GREEN]

clear 함수 호출시 모든 배열의 요소가 지워지는 것을 확인했습니다.

Commit:
:fire: Refactor: 불필요한 extension 삭제.

불필요한 extension String, extension Double를 삭제하였습니다.

wonbi92 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,38 @@
+
+protocol CalculateItem { }
+
+struct CalculatorItemQueue<Element: CalculateItem> where Element: Equatable {
```
`where`절을 통해 `Equatable`로 제한하신 이유를 알 수 있을까요? 🤔

devKobe24 replied:
저 코드를 구현할 당시에 Element의 대소비교를 가능하게 하려고 Equatable로 제한을 두었습니다.🙋‍♂️

wonbi92 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,38 @@
+struct CalculatorItemQueue<Element: CalculateItem> where Element: Equatable {
+	private var elements: [Element] = []
+	
+	func getElementsForTest() -> [Element] {
+		return elements
+	}
```
테스트를 위한 메서드를 작성하셨군요..!
해당 로직 자체는 테스트에서만 필요하다는 생각이 들긴 합니다만, 주객전도라는 생각도 듭니다.
테스트의 방법을 바꿔보시거나 해당 로직을 테스트 객체 안으로 옮기시는것도 좋아보여요.
아니면 Test Double을 이용해 Mock객체를 만들어서 해당 객체에만 이 로직을 옮기는 것도 방법이 될 수 있겠네요!

테스트를 위한 메서드나 프로퍼티가 실제 객체에 존재하는 거 자체가 조금 어색해 보입니다. 
이 부분은 다른 방법을 고려해볼 필요가 있겠네요.

devKobe24 replied:
현재 Test Double을 공부하고 Mock 객체를 따로 만들어 테스트에 사용하였습니다 🙋‍♂️

wonbi92 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,38 @@
+	}
+	
+	mutating func dequeue() -> Element? {
+		if elements != [] {
+			return elements.removeFirst()
+		} else {
+			return nil
+		}
```
사바사겠지만, 저였다면 요런식으로 표현하는게 좀 더 간단해 보이네요! 어떻게 생각하세요?
```suggestion
		elements.isEmpty ? nil : elements.removeFirst()
```

devKobe24 replied:
원비께서 피드백주신 삼항 연산자가 훨씬 깔끔하고 가독성이 좋습니다. 가독성이 좋은 코드를 만들기 위해 노력하겠습니다. 🙋‍♂️

wonbi92 reviewed (commented):
@devKobe24 
리팩토링 해주신 부분들 잘 봤습니다! 고생하셨어요! 👍

추가적인 코맨트를 드렸으니 참고해보시기 바랍니다. 

---

현재 큐를 구현함에 있어 링크드 리스트에 대한 공부가 부족해 배열을 사용하신 부분은 이해가 됩니다. 다만, 큐를 구현하는 방법은 배열을 사용하는 것과 링크드 리스트를 사용하는 방법 외에도 몇가지가 더 있는데요. 해당 부분에 대해 고려하신 것 같지 않아서 이 부분은 조금 아쉽네요..! 
배열과 링크드 리스트 외에 다른 방법들도 한번 고려해보시고, 최종적으로 어떤 방법을 선택하실지 좀더 고민해보시길 바랍니다!

Commit:
:memo: Add: update README.md

계산기 프로젝트 STEP 1 README를 작성 후 업로드 합니다.

Commit:
:sparkles: Feat: Node 파일 생성, Node 클래스 구현.

제네릭 타입을 활용하여 Node 클래스를 구현했습니다.

Commit:
:sparkles: Feat: LinkedList 파일 생성.

LikedList 클래스를 만들기 위한 파일을 생성했습니다.

Commit:
:sparkles: Feat: Node, LinkedList 파일 생성, Node 클래스 구현.

Node, LinkedList 파일 생성, Node 클래스 구현.

Commit:
:sparkles: Feat: LinkedList 프로토콜, LinkedList 클래스 구현.

LinkedList 프로토콜을 구현하고 그 프로토콜을 채택하여 LinkedList 클래스를 구현하였습니다.

Commit:
:recycle: Refactor: LinkedListProtocol에 isEmpty 속성 추가.

LinkedListProtocol에 Bool 타입의 isEmpty 속성을 추가해줬습니다.

Commit:
:recycle: Refactor: Node를 extension.

Node를 Extension한 후 Node가 CalculateItem 프로토콜을 채택하게 하였습니다.

Commit:
:recycle: Refactor: LinkedListProtocol 함수명 변경.

LinkedListProtocol에 있었던 enqueue와 dequeue를 각각 append와 removeFirst로 함수명을 변경했습니다.

Commit:
:recycle: Refactor: removeFirst 함수 로직 변경.

removeFirst 내부 함수 로직을 변경하고 반환값을 주었습니다.

Commit:
:sparkles: Feat: Test Double를 위한 CalculateItemProtocol 구현

Test Double를 위한 빈 프로토콜인 CalculateItemProtocol을 구현했습니다.

Commit:
:sparkles: Feat: DummyLinkedList, MockCalculatorItemQueue 구현.

Test Double를 위해 DummyLinkedList와 MockCalculatorItemQueue를 구현했습니다.

Commit:
:recycle: Refactor: private 제어 접근자 삭제.

head, tail 프로퍼티에 private 을 삭제해주었습니다. CalculateItemQueue에서 접근이 불가해서 삭제했습니다.

Commit:
:recycle: Refactor: CalculatorItemQueue 단방향 리스트로 수정.

CalculatorItemQueue를 단방향 리스트로 전면 수정하였습니다.

Commit:
:recycle: Refactor: 코드 전면 수정.

CalculatorItemQueue, LinkedList 수정, CalculateItemProtocl, MockCalculatorItemQueue 생성

Commit:
:white_check_mark: Test: Mock Object와 Dummy Data 활용한 테스트 성공[GREEN]

Commit:
:recycle: Refactor: 접근 제어자 설정.

Public과 fileprivate를 각각 설정해주었습니다.

Commit:
:test_tube: Test: enqueue 확인 테스트 [RED]

enqueue 호출시 데이터가 쌓이는지 확인하는 테스트를 해본 결과 실패했습니다.

Commit:
:white_check_mark: Test: enqueue 호출시 데이터가 쌓이는 테스트[GREEN]

enqueue 호출시 데이터가 쌓여 nil이 아님을 확인했습니다.

Commit:
:sparkles: Feat: MockNode 클래스 생성.

Mock Object인 MockNode 클래스를 생성하였습니다.

Commit:
:recycle: Refactor: Node타입을 MockNode<T> 타입으로 변경.

Node<T> 타입이였던 프로퍼티나 Node(value:)가 할당된 프로퍼티들을 모두 MockNode<T> 또는 MockNode(value:)로 변경해주었습니다.

Commit:
:recycle: Refactor: CalculateItemProtocol에 Equatable을 채택.

Mock Object인 CalculateItemProtocol에 Equatable을 비교 연산이 가능하도록 채택해주었습니다.

Commit:
:white_check_mark: Test: MockNode의 Value값 비교 테스트[GREEN]

MockNode의 value 값을 서로 다르게 해놓은 뒤 비교하여 다른 것을 확인했습니다.

Commit:
refactor: 테스트 코드 성공, MockNode 생성, Equatable 채택, MockNode 타입으로 변경.

Commit:
:recycle: Refactor: CalculateItem 프로토콜 Equatable 채택.

CalculateItem 프로토콜을 준수하는 객체들은 대소비교가 가능하게 만들기 위해 Equatable 프로토콜을 채택했습니다.

Commit:
:recycle: Refactor: LinkedList의 T가 CalculateItem 채택.

LinkedList의 제네릭 T 타입이 CalculateItem을 채택함으로써 CalculatorItemQueue에서 다루는 요소로서 가능하게 만들었습니다.

Commit:
:recycle: Refactor: Node 클래스와 T 타입 Equatable 채택.

Node 클래스가 Equatable을 채택함으로써 Equtable을 준수하도록하였습니다. 그래서 대소비교가 가능하게 하였습니다. T 타입 역시 Equatable을 채택했습니다.

Commit:
:fire: Remove: 불필요한 코드 삭제.

테스트 케이스에서 불필요한 코드를 삭제했습니다.

Commit:
:recycle: Refactor: 접근 제어자 변경, MockLinkedListProtocol 생성.

메서드의 접근제어가자 fileprivate에서 internal로 변경되었습니다. MockLinkedListProtocol을 생성하고 DummyLinkedList가 채택하고 준수하였습니다.

Commit:
:recycle: Refactor: 접근 제어자 삭제.

DummyLinkedList의 head와 tail에 fileprivate을 삭제했습니다.

Commit:
:recycle: Refactor: Mock enqueue Test code 수정.

enqueue 호출시 dummyList에 데이터가 쌓인다 라는 테스트 케이스의 input과 result 코드 그리고 XCTAssertEqual로 변경하여 테스트했습니다.

Commit:
:white_check_mark: Test: dequeue시 데이터 삭제 테스트[GREEN]

Mock dequeue 호출시 데이터가 dummyList에서 삭제되는지 확인하는 테스트를 실행했으며 성공했습니다.

Commit:
:white_check_mark: Test: isEmpty 호출시 true 반환 테스트[GREEN]

빈 dummyList일 경우 isEmpty를 호출하면 true를 반환하는 테스트를 실행했고 성공했습니다.

Commit:
:white_check_mark: Test: isEmpty 호출시 false 반환 테스트[GREEN]

isEmpty 호출시 dummyList에 데이터가 들어있을 경우 false를 반환하는 테스트를 실행하였고 성공했습니다.

Commit:
:white_check_mark: Test: front 호출시 head 값 가져오는 테스트[GREEN]

dummyList의 head에 데이터가 들어있을 경우에는 그 head 값을 가져오는 테스트를 시행하였고 성공했습니다.

Commit:
:white_check_mark: Test: front 호출시 nil값 반환 테스트[GREEN]

dummyList의 head에 데이터가 없을 경우에 front를 호출하면  nil을 반환하는 테스트를 실행하였고 성공했습니다.

Commit:
:recycle: Refactor: DummyLinkedList에서 MockLinkedList로 변경.

DummyLinkedList에서 MockLinkedList로 이름을 변경하고 새롭게 파일을 만들어 분리하였습니다.

Commit:
:recycle: Refactor: MockNode에서 DummyNode로 변경.

MockNode에서 DummyNode로 네이밍을 변경하였습니다.

wonbi92 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,35 @@
+protocol CalculateItem: Equatable { }
+
+struct CalculatorItemQueue<T: CalculateItem> {
+	private var list: LinkedList<T>?
```
큐 내부의 리스트가 꼭 옵셔널일 필요가 있을까요? 🤔 리스트가 옵셔널일 경우 어떻게 되나요? 한번 고민해보시기 바랍니다!

wonbi92 commented on Calculator/Calculator/Model/LinkedList.swift:
```diff
@@ -0,0 +1,41 @@
+//
+
+protocol LinkedListProtocol {
+	associatedtype T: Any
```
`associatedtype T`가 `Any`를 채택한 이유가 무엇일까요?

wonbi92 commented on Calculator/TestDouble/MockCalculatorItemQueue.swift:
```diff
@@ -0,0 +1,40 @@
+
+extension Int: CalculateItemProtocol { }
+
+protocol MockLinkedListProtocol {
+	associatedtype T: Any
+	var isEmpty: Bool { get }
+	func append(value: T)
+	func removeFirst() -> T?
+}
```
링크드리스트를 프로토콜로 한번 추상화를 하였는데 다시 목객체를 추상화 하고 있네요..! 이러면 테스트 링크드 리스트와 실제 링크드리스트는 완전히 별개의 객체로 되어있는거 같아요..! TestDouble을 활용한 Mock객체를 만드는 이유에 대해 조금 더 고민해보시면 좋을 것 같습니다.

wonbi92 reviewed (approved):
@devKobe24 
안녕하세요 코비!
잘 수정해주셨어요! 고민하신 흔적이 보여서 좋았습니다! 👍👍

약간 고민하실만한 부분들이 있지만 다음 스탭 진행하시면서 하셔도 충분할거 같아요! 
코맨트 읽어보시고 다음 스탭 진행하시면서 같이 적용시켜 보시기 바랍니다!
